File: ios/Amrosa/Amrosa/UI/Shopping/ShoppingAggregator.swift
import Foundation

/// One combined line on the shopping list. `key` is the persisted checked-state key.
struct ShoppingLine: Identifiable {
    let key: String
    let name: String
    let quantity: String
    let note: String?

    var id: String { key }
}

/// Combines a recipe's ingredients into a flat shopping list: ingredients with the same
/// (normalized) name are merged into one line with their quantities summed. Summing happens
/// per unit within the active `UnitMode` — Metric mode reduces most items to clean g/ml; when
/// units in a group don't match they're joined with " + ". Non-numeric amounts ("to taste")
/// are carried through as-is. Author shopping notes are aggregated onto the line.
enum ShoppingAggregator {

    /// Mirror of the backend `normalizeKey` so learned keys line up conceptually.
    static func normalizeKey(_ name: String) -> String {
        let lowered = name.lowercased()
        let cleaned = lowered.map { ch -> Character in
            (ch.isLetter && ch.isASCII) || ch.isNumber || ch == " " ? ch : " "
        }
        return String(cleaned)
            .components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
    }

    private static func valueUnit(for ing: IngredientModel, mode: UnitMode) -> (Double?, String?) {
        switch mode {
        case .metric:
            if let v = ing.quantityValueMetric { return (v, ing.quantityUnitMetric) }
            return (ing.quantityValue, ing.quantityUnit)
        case .imperial:
            if let v = ing.quantityValueImperial { return (v, ing.quantityUnitImperial) }
            return (ing.quantityValue, ing.quantityUnit)
        case .original:
            return (ing.quantityValue, ing.quantityUnit)
        }
    }

    static func build(
        ingredients: [IngredientModel],
        scaleFactor: Double,
        unitMode: UnitMode
    ) -> [ShoppingLine] {
        // Preserve first-seen order of ingredient names.
        var order: [String] = []
        var groups: [String: [IngredientModel]] = [:]
        for ing in ingredients {
            let key = normalizeKey(ing.name)
            if key.isEmpty { continue }
            if groups[key] == nil { order.append(key) }
            groups[key, default: []].append(ing)
        }

        return order.compactMap { key in
            guard let members = groups[key] else { return nil }

            // Sum numeric amounts bucketed by unit; carry non-numeric amounts as display text.
            var bucketOrder: [String] = []
            var buckets: [String: (sum: Double, unit: String?)] = [:]
            var nonNumeric: [String] = []

            for ing in members {
                let (value, unit) = valueUnit(for: ing, mode: unitMode)
                if let value = value {
                    let unitKey = (unit ?? "").lowercased().trimmingCharacters(in: .whitespaces)
                    if buckets[unitKey] == nil { bucketOrder.append(unitKey) }
                    let existing = buckets[unitKey]
                    buckets[unitKey] = ((existing?.sum ?? 0) + value, existing?.unit ?? unit)
                } else if let display = ing.quantityDisplay, !display.isEmpty, !nonNumeric.contains(display) {
                    nonNumeric.append(display)
                }
            }

            let parts = bucketOrder.compactMap { unitKey -> String? in
                guard let bucket = buckets[unitKey] else { return nil }
                return QuantityScaler.scale(
                    quantityValue: bucket.sum, quantityUnit: bucket.unit,
                    quantityDisplay: nil, scale: scaleFactor
                )
            } + nonNumeric

            let notes = members
                .compactMap { $0.shoppingNote?.trimmingCharacters(in: .whitespaces) }
                .filter { !$0.isEmpty }
            var seenNotes = Set<String>()
            let note = notes.filter { seenNotes.insert($0).inserted }.joined(separator: "; ")

            return ShoppingLine(
                key: key,
                name: members.first?.name ?? key,
                quantity: parts.joined(separator: " + "),
                note: note.isEmpty ? nil : note
            )
        }
    }
}

File: android/app/src/main/java/com/aerion/chefsjournal/ui/shopping/ShoppingAggregator.kt
package com.aerion.chefsjournal.ui.shopping

import com.aerion.chefsjournal.domain.model.Ingredient
import com.aerion.chefsjournal.ui.util.QuantityScaler
import com.aerion.chefsjournal.ui.util.UnitMode

/** One combined line on the shopping list. [key] is the persisted checked-state key. */
data class ShoppingLine(
    val key: String,
    val name: String,
    val quantity: String,
    val note: String?,
)

/**
 * Combines a recipe's ingredients into a flat shopping list: ingredients with the same
 * (normalized) name are merged into one line with their quantities summed. Summing happens
 * per unit within the active [UnitMode] — Metric mode reduces most items to clean g/ml; when
 * units in a group don't match they're joined with " + ". Non-numeric amounts ("to taste")
 * are carried through as-is. Author shopping notes are aggregated onto the line.
 */
object ShoppingAggregator {

    /** Mirror of the backend `normalizeKey` so learned keys line up conceptually. */
    fun normalizeKey(name: String): String =
        name.lowercase()
            .replace(Regex("[^a-z0-9 ]"), " ")
            .replace(Regex("\\s+"), " ")
            .trim()

    private fun valueUnitFor(ing: Ingredient, mode: UnitMode): Pair<Double?, String?> =
        when (mode) {
            UnitMode.METRIC ->
                if (ing.quantityValueMetric != null) ing.quantityValueMetric to ing.quantityUnitMetric
                else ing.quantityValue to ing.quantityUnit
            UnitMode.IMPERIAL ->
                if (ing.quantityValueImperial != null) ing.quantityValueImperial to ing.quantityUnitImperial
                else ing.quantityValue to ing.quantityUnit
            UnitMode.ORIGINAL -> ing.quantityValue to ing.quantityUnit
        }

    fun build(
        ingredients: List<Ingredient>,
        scaleFactor: Double,
        unitMode: UnitMode,
    ): List<ShoppingLine> {
        // Preserve first-seen order of ingredient names.
        val groups = LinkedHashMap<String, MutableList<Ingredient>>()
        for (ing in ingredients) {
            val key = normalizeKey(ing.name)
            if (key.isEmpty()) continue
            groups.getOrPut(key) { mutableListOf() }.add(ing)
        }

        return groups.map { (key, members) ->
            // Sum numeric amounts bucketed by unit; carry non-numeric amounts as display text.
            val buckets = LinkedHashMap<String, Pair<Double, String?>>() // unitKey -> (sum, displayUnit)
            val nonNumeric = LinkedHashSet<String>()

            for (ing in members) {
                val (value, unit) = valueUnitFor(ing, unitMode)
                if (value != null) {
                    val unitKey = (unit ?: "").lowercase().trim()
                    val existing = buckets[unitKey]
                    buckets[unitKey] = ((existing?.first ?: 0.0) + value) to (existing?.second ?: unit)
                } else {
                    ing.quantityDisplay?.takeIf { it.isNotBlank() }?.let { nonNumeric.add(it) }
                }
            }

            val parts = buckets.map { (_, sumUnit) ->
                QuantityScaler.scale(sumUnit.first, sumUnit.second, null, scaleFactor)
            } + nonNumeric.toList()

            val note = members.mapNotNull { it.shoppingNote?.trim()?.ifBlank { null } }
                .distinct()
                .joinToString("; ")
                .ifBlank { null }

            ShoppingLine(
                key = key,
                name = members.first().name,
                quantity = parts.joinToString(" + "),
                note = note,
            )
        }
    }
}
